packages/contracts/contracts/GenArt721CoreV4.sol

```diff
@@ -0,0 +1,2010 @@
+     * determined by this contract's current Admin ACL contract. Expected use
+     * cases include minter contracts checking if caller is allowed to call
+     * admin-gated functions on minter contracts.
+     * @param _sender Address of the sender calling function with selector
+     * `_selector` on contract `_contract`.
+     * @param _contract Address of the contract being called by `_sender`.
+     * @param _selector Function selector of the function being called by
+     * `_sender`.
+     * @return bool Whether `_sender` is allowed to call function with selector
+     * `_selector` on contract `_contract`.
+     * @dev assumes the Admin ACL contract is the owner of this contract, which
+     * is expected to always be true.
+     * @dev adminACLContract is expected to either be null address (if owner
+     * has renounced ownership), or conform to IAdminACLV0 interface. Check for
+     * null address first to avoid revert when admin has renounced ownership.
+     */
+    function adminACLAllowed(
+        address _sender,
+        address _contract,
+        bytes4 _selector
+    ) public returns (bool) {
+        return
+            owner() != address(0) &&
+            adminACLContract.allowed(_sender, _contract, _selector);
+    }
+
+    /**
+     * @notice Returns contract owner. Set to deployer's address by default on
+     * contract deployment.
+     * @return address Address of contract owner.
+     * @dev ref: https://docs.openzeppelin.com/contracts/4.x/api/access#Ownable
+     * @dev owner role was called `admin` prior to V3 core contract
+     */
+    function owner()
+        public
+        view
+        override(Ownable, IGenArt721CoreContractV3_Base)
+        returns (address)
+    {
+        return Ownable.owner();
+    }
+
+    /**
+     * @notice Gets token URI for token ID `_tokenId`.
+     * @param _creatorCore Address of creatorCore contract.
+     * @param _tokenId Token ID to be queried.
+     * @return string URI of token ID `_tokenId`.
+     * @dev token URIs are the concatenation of the project base URI and the
+     * token ID.
+     */
+    function tokenURI(
+        address _creatorCore,
+        uint256 _tokenId
+    ) public view override returns (string memory) {
+        _onlyValidTokenId(_tokenId);
+        require(
+            _creatorCore ==
+                projects[tokenIdToProjectId(_tokenId)].artistContractAddress,
+            "Invalid creatorCore"
+        );
+        string memory _projectBaseURI = projects[tokenIdToProjectId(_tokenId)]
+            .projectBaseURI;
+        return string.concat(_projectBaseURI, _tokenId.toString());
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(
+        bytes4 interfaceId
+    ) public view virtual override returns (bool) {
+        return
+            interfaceId == type(IManifold).interfaceId ||
+            interfaceId == type(IManifoldCreatorExtensionURI).interfaceId ||
```

Comment:
It now implements the `IManifoldCreatorExtensionURI` which tells the artist's contract that it's OK to fetch `tokenURI` info from here.